### DISCUSSION TITLE
什麼情況下 request 會回傳 error ?

### DISCUSSION THREAD

**Discussion by neniz77** [Lidemy/forum#120] [程式導師實驗計畫]
### 作業名稱
week4 hw3 周遊列國

### 碰到的問題
我目前的理解是，在 request 第二個參數的函式當中，如果沒有 error ，才有辦法讀取 response 和 body。如果 body 是 JSON 格式字串，才有辦法 parse 成物件，然後才能去讀取 status code，確認 status code 是不是 404。所以在能夠取得 http 狀態碼之前，還有兩個關卡需要確認。我想了解的是：

1. error 會在什麼樣的狀況下發生？
2. 沒有 error，但在 try {data = JSON.parse(body)} 那邊發生錯誤的情形，會在什麼樣的狀況下發生？

### 嘗試過的解決方法
自己測試了一下，有測到：
1. 如果 domain name 打錯，會出現 error。
2. domain name 正確，但 path 是一個不存在的路徑的話， body 就不會是 JSON 格式字串。

好像找到了一些現象但自己不知道該怎麼解釋這些現象？用關鍵字 'http response error' 去 google，都只能查到狀態碼相關資料

### 相關程式碼
```javascript
request(`${API_ENDPOINT}/name/${name}`,
  (error, response, body) => {
    if (error) {
      return console.log(`抓取失敗，錯誤訊息如下：${error}`)
    }
    let data
    try {
      data = JSON.parse(body)
    }catch (e) {
      console.log(`無法讀取資料，錯誤訊息如下：${e}`)
      return
    }
    if (data.status === 404) {
      return console.log('找不到國家資訊')
    }
    for (let i = 0; i < data.length; i++) {
      console.log(`============
        國家：${data[i].name}
        首都：${data[i].capital}
        貨幣：${data[i].currencies[0].code}
        國碼：${data[i].callingCodes}
        `)
    }
  })
```

**Comment by cmtilo** [comment]
就我目前理解是這個樣子...
錯誤有3種

1.library 的錯誤（request() 的錯誤）： 對 library 來說是否為錯誤要看不同的 library，以作業案例來說，對 request 而言，回傳狀態碼錯誤並不認為是錯誤，連 request 都發不出去才認為是錯誤。
2.JSON.parse() 用 try catch 去接到的錯誤： 上面提到 request 判斷只要有拿到 response 就算成功，但對我們來說拿到資料格式不對就是一種錯誤，所以用 try catch 包住的錯誤是為了以防資料格式不是 JSON 的錯誤。簡言之，JSON.parse() 用 try catch 去接代表：我們希望判斷資料格式如果不是 JSON 格式就跳錯誤訊息。 
3.response.statusCode狀態碼錯誤


所以你的第2個問題是資料格式不對我們就希望程式判斷輸出錯誤，那個path不一樣了就不是原本給的資料了

**Reply by allenliaooooooo** [reply]
如果資料格式不是 json 格式的話，程式會自己丟出錯誤，因為他不是 json 格式你卻要用 `JSON.parse()` 去解析他
try...catch 應該只是把這個錯誤接起來，你可以在 catch 裡面自己寫接到這個錯誤要做什麼事，你甚至可以讓程式碼繼續執行，不理會這個錯誤

而 status code 只是一個代碼，代表回傳資料的狀態，只要有 response 一定會有 status code
而我們可以透過 status code 去大概知道 request 或 response 有沒有問題
